src/Waives.Pipelines/ConcurrentPipelineObserver.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace Waives.Pipelines
+{
+    internal class ConcurrentPipelineObserver : IObserver<Document>, IDisposable
+    {
+        private readonly IDocumentProcessor _documentProcessor;
+        private readonly Action _onPipelineCompleted;
+        private bool _sourceComplete;
+
+        private readonly WorkPool _workPool;
+
+        internal ConcurrentPipelineObserver(
+            IDocumentProcessor documentProcessor,
+            Action onPipelineCompleted,
+            int maxConcurrency)
+        {
+            _documentProcessor = documentProcessor ?? throw new ArgumentNullException(nameof(documentProcessor));
+            _onPipelineCompleted = onPipelineCompleted ?? throw new ArgumentNullException(nameof(onPipelineCompleted));
+            _workPool = new WorkPool(maxConcurrency);
+        }
+
+        public void OnCompleted()
+        {
+            _sourceComplete = true;
+            _ = TryCompletionHandler();
```

Comment:
✋ This doesn't look right to me. We're deliberately not awaiting this `Task`, so I think any exception thrown by the `_onPipelineCompleted()` action will be swallowed?